In this project, we will focus on sensor data. The dataset contains accelerometer data from cell phones. Accelerometer helps measure the speed and acceleration of a cell phone's movement. Each row represents a single measurement (captured on a timeline). There are a total of 20 time steps (columns). This is a multiclass classification task: predict what type of transportation each measurement (i.e., row) represents based on the accelerometer data. 

## Description of Variables

I will use the **movement.csv** data set for this. Each row represents a single measurement. Columns labeled as 1 from 20 are the time steps on the timeline (there are 20 time steps, each time step has only one measurement). 


## Goal

Using the data set **movement.csv** to predict the column called **Target**. The input variables are columns labeled as **1 to 20** in time steps.

# Read and Prepare the Data 

In [1]:
# Common imports
import numpy as np
import tensorflow as tf
from tensorflow import keras
import pandas as pd

np.random.seed(123)

In [2]:
data = pd.read_csv("movement.csv")

In [3]:
data.shape

(118, 21)

In [4]:
data.head()

1         2         3         4         5         6         7  \
0  1.179784  1.179784  0.810629  0.810629  1.041816  1.041816  1.041816   
1  1.115912  0.860983  0.860983  0.860983  0.860983  1.020423  1.020423   
2  0.572300  0.147040  0.147040  0.147040  0.147040  0.147040  0.147040   
3  1.128633  1.128633  1.128633  1.128633  1.128633  3.181596  4.012386   
4  0.548065  0.548065  0.548065  1.441688  0.631261  0.631261  0.631261   

          8         9        10  ...        12        13        14        15  \
0  0.453604  0.453604  0.483920  ...  0.250571  0.250571  0.250571  0.250571   
1  1.333723  1.333723  1.333723  ...  1.582763  0.936744  0.936744  0.936744   
2  0.147040  0.147040  2.662993  ...  2.662993  2.662993  2.662993  1.449779   
3  4.012386  1.349989  1.266019  ...  1.266019  1.266019  0.492464  0.710132   
4  9.258070  1.495908  0.723835  ...  8.073005  8.073005  8.073005  8.073005   

         16        17        18        19        20  Target  
0  0.250571  0.250571  0.167502  0.167502  0.167502     Bus  
1  0.936744  1.412754  3.283429  3.283429  3.283429     Bus  
2  1.449779  1.147295  0.978355  0.978355  0.978355     Bus  
3  0.710132  0.251398  0.251398  1.347456  1.347456     Bus  
4  1.124158  0.399042  0.399042  0.399042  0.561521     Car  

[5 rows x 21 columns]

In [5]:
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(data, test_size=0.3)

In [6]:
train_y = train_set[['Target']]
test_y = test_set[['Target']]

train_x = train_set.drop(['Target'], axis=1)
test_x = test_set.drop(['Target'], axis=1)

In [7]:
# Identify the numerical columns
numeric_columns = train_x.select_dtypes(include=[np.number]).columns.to_list()

# Identify the categorical columns
categorical_columns = train_x.select_dtypes('object').columns.to_list()

In [8]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

from sklearn.preprocessing import FunctionTransformer


In [9]:
numeric_transformer = Pipeline(steps=[
                ('imputer', SimpleImputer(strategy='mean')),
                ('scaler', StandardScaler())])

In [10]:
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='unknown')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

In [11]:
preprocessor = ColumnTransformer([
        ('num', numeric_transformer, numeric_columns),
        ('cat', categorical_transformer, categorical_columns)
],
        remainder='drop')

#passtrough is an optional step. You don't have to use it.

# Keras needs Ordinal target values for classification

In [12]:
from sklearn.preprocessing import OrdinalEncoder

ord_enc = OrdinalEncoder()
train_target_2d = train_y.values.reshape(-1, 1)

train_y = ord_enc.fit_transform(train_target_2d)

train_y

array([[4.],
       [3.],
       [1.],
       [2.],
       [1.],
       [0.],
       [4.],
       [1.],
       [1.],
       [1.],
       [2.],
       [2.],
       [2.],
       [4.],
       [3.],
       [4.],
       [3.],
       [4.],
       [1.],
       [0.],
       [3.],
       [2.],
       [2.],
       [4.],
       [4.],
       [3.],
       [1.],
       [3.],
       [4.],
       [3.],
       [4.],
       [4.],
       [3.],
       [3.],
       [3.],
       [2.],
       [4.],
       [2.],
       [4.],
       [0.],
       [0.],
       [1.],
       [2.],
       [2.],
       [4.],
       [1.],
       [1.],
       [2.],
       [4.],
       [0.],
       [4.],
       [1.],
       [1.],
       [3.],
       [0.],
       [0.],
       [2.],
       [4.],
       [0.],
       [3.],
       [3.],
       [4.],
       [2.],
       [4.],
       [4.],
       [4.],
       [2.],
       [4.],
       [1.],
       [2.],
       [4.],
       [4.],
       [4.],
       [4.],
       [3.],
       [3.],
       [3.],

In [13]:
test_target_2d = test_y.values.reshape(-1, 1)

test_y = ord_enc.transform(test_target_2d)

test_y

array([[1.],
       [4.],
       [4.],
       [3.],
       [4.],
       [1.],
       [4.],
       [4.],
       [4.],
       [4.],
       [2.],
       [1.],
       [1.],
       [0.],
       [3.],
       [1.],
       [1.],
       [0.],
       [0.],
       [2.],
       [4.],
       [1.],
       [3.],
       [3.],
       [4.],
       [4.],
       [3.],
       [4.],
       [2.],
       [4.],
       [4.],
       [3.],
       [4.],
       [1.],
       [4.],
       [0.]])

## Data Transformation

In [14]:
#Target variables need to be an array with integer type
train_y = np.array(train_y)
test_y = np.array(test_y)


In [15]:
train_y = train_y.astype(np.int32)
test_y = test_y.astype(np.int32)

In [16]:

train_y = train_y.flatten()


In [17]:
test_y = test_y.flatten()

In [18]:
train_y[0:10]


array([4, 3, 1, 2, 1, 0, 4, 1, 1, 1])

In [19]:
#Convert input variables to a 2-D array with float data type
train_x= np.array(train_x)
test_x= np.array(test_x)

train_x = train_x.astype(np.float32)
test_x = test_x.astype(np.float32)

In [20]:
train_x

array([[ 3.574058  ,  3.574058  ,  3.574058  , ...,  3.1191604 ,
         3.1191604 ,  3.3081055 ],
       [ 0.6157446 ,  1.134273  ,  0.5657411 , ...,  0.12147009,
         0.0679978 ,  0.0403792 ],
       [ 4.9917274 ,  4.9917274 ,  4.9917274 , ...,  0.7166513 ,
         0.7166513 ,  0.7166513 ],
       ...,
       [ 8.132629  , 11.646927  , 11.960547  , ...,  6.5444446 ,
         6.5539155 , 12.310273  ],
       [ 2.852925  ,  5.4435086 ,  5.4435086 , ...,  3.9759524 ,
         5.9093347 ,  5.9093347 ],
       [ 2.3752983 ,  3.7121398 ,  3.7121398 , ...,  1.5666957 ,
         8.141131  ,  6.1379995 ]], dtype=float32)

In [21]:
train_y

array([4, 3, 1, 2, 1, 0, 4, 1, 1, 1, 2, 2, 2, 4, 3, 4, 3, 4, 1, 0, 3, 2,
       2, 4, 4, 3, 1, 3, 4, 3, 4, 4, 3, 3, 3, 2, 4, 2, 4, 0, 0, 1, 2, 2,
       4, 1, 1, 2, 4, 0, 4, 1, 1, 3, 0, 0, 2, 4, 0, 3, 3, 4, 2, 4, 4, 4,
       2, 4, 1, 2, 4, 4, 4, 4, 3, 3, 3, 3, 0, 4, 4, 4])

In [22]:
#Keras expects a different input format:
#Data needs to have 3 dimensions

train_x = np.reshape(train_x, (train_x.shape[0], train_x.shape[1], 1))
test_x = np.reshape(test_x, (test_x.shape[0], test_x.shape[1], 1))



In [23]:
train_x.shape, train_y.shape

((82, 20, 1), (82,))

In [24]:
train_x

array([[[ 3.574058  ],
        [ 3.574058  ],
        [ 3.574058  ],
        ...,
        [ 3.1191604 ],
        [ 3.1191604 ],
        [ 3.3081055 ]],

       [[ 0.6157446 ],
        [ 1.134273  ],
        [ 0.5657411 ],
        ...,
        [ 0.12147009],
        [ 0.0679978 ],
        [ 0.0403792 ]],

       [[ 4.9917274 ],
        [ 4.9917274 ],
        [ 4.9917274 ],
        ...,
        [ 0.7166513 ],
        [ 0.7166513 ],
        [ 0.7166513 ]],

       ...,

       [[ 8.132629  ],
        [11.646927  ],
        [11.960547  ],
        ...,
        [ 6.5444446 ],
        [ 6.5539155 ],
        [12.310273  ]],

       [[ 2.852925  ],
        [ 5.4435086 ],
        [ 5.4435086 ],
        ...,
        [ 3.9759524 ],
        [ 5.9093347 ],
        [ 5.9093347 ]],

       [[ 2.3752983 ],
        [ 3.7121398 ],
        [ 3.7121398 ],
        ...,
        [ 1.5666957 ],
        [ 8.141131  ],
        [ 6.1379995 ]]], dtype=float32)

In [25]:
train_y

array([4, 3, 1, 2, 1, 0, 4, 1, 1, 1, 2, 2, 2, 4, 3, 4, 3, 4, 1, 0, 3, 2,
       2, 4, 4, 3, 1, 3, 4, 3, 4, 4, 3, 3, 3, 2, 4, 2, 4, 0, 0, 1, 2, 2,
       4, 1, 1, 2, 4, 0, 4, 1, 1, 3, 0, 0, 2, 4, 0, 3, 3, 4, 2, 4, 4, 4,
       2, 4, 1, 2, 4, 4, 4, 4, 3, 3, 3, 3, 0, 4, 4, 4])

# Baseline

In [26]:
from sklearn.dummy import DummyClassifier

dummy_clf = DummyClassifier(strategy="most_frequent")

dummy_clf.fit(train_x, train_y)

DummyClassifier(strategy='most_frequent')

In [27]:
from sklearn.metrics import accuracy_score

In [28]:
#Baseline Train Accuracy
dummy_train_pred = dummy_clf.predict(train_x)

baseline_train_acc = accuracy_score(train_y, dummy_train_pred)

print('Baseline Train Accuracy: {}' .format(baseline_train_acc))

Baseline Train Accuracy: 0.34146341463414637


In [29]:
#Baseline Test Accuracy
dummy_test_pred = dummy_clf.predict(test_x)

baseline_test_acc = accuracy_score(test_y, dummy_test_pred)

print('Baseline Test Accuracy: {}' .format(baseline_test_acc))

Baseline Test Accuracy: 0.4166666666666667


In [30]:
train_x.shape

(82, 20, 1)

In [31]:
train_y.shape

(82,)

In [32]:
test_x.shape

(36, 20, 1)

In [33]:
test_y.shape

(36,)

# Cross-sectional (i.e., a regular) Neural Network model using Keras (with only one hidden layer)

In [34]:
np.random.seed(42)
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[20, 1]),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(5, activation='softmax')
    
])

C:\Users\neera\anaconda3\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [35]:
tf.random.set_seed(42)
np.random.seed(42)

optimizer = tf.keras.optimizers.Nadam(learning_rate=0.1)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=7,
                    validation_data=(test_x, test_y))

Epoch 1/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 3s 181ms/step - accuracy: 0.3257 - loss: 3.0395 - val_accuracy: 0.4444 - val_loss: 1.3907
Epoch 2/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.4219 - loss: 1.3018 - val_accuracy: 0.5278 - val_loss: 1.1170
Epoch 3/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step - accuracy: 0.4592 - loss: 1.1801 - val_accuracy: 0.5833 - val_loss: 1.0050
Epoch 4/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5053 - loss: 1.0205 - val_accuracy: 0.5833 - val_loss: 0.9769
Epoch 5/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - accuracy: 0.5649 - loss: 0.9383 - val_accuracy: 0.5833 - val_loss: 1.0741
Epoch 6/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 51ms/step - accuracy: 0.5571 - loss: 0.9727 - val_accuracy: 0.6111 - val_loss: 0.9041
Epoch 7/7
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step - accuracy: 0.5510 - loss: 0.8597 - val_accuracy: 0.6111 - val_loss: 0.8950


In [36]:
# evaluate the model

scores = model.evaluate(train_x, train_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.738736629486084, 0.707317054271698]

In [37]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.8949598670005798, 0.6111111044883728]

In [38]:
# extract the accuracy from model.evaluate

print("%s of neural network with one hidden layer : %.2f" % (model.metrics_names[0], scores[0]))
print("%s of neural network with one hidden layer : %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss of neural network with one hidden layer : 0.89
compile_metrics of neural network with one hidden layer : 61.11%


# Deep cross-sectional (i.e., regular) Neural Network model using Keras (with two or more hidden layers)

In [39]:
np.random.seed(42)
model = keras.models.Sequential()

model.add(keras.layers.Input(shape=(20,)))
model.add(keras.layers.Dense(30, activation='relu'))
model.add(keras.layers.Dense(20, activation='elu'))
model.add(keras.layers.Dense(10, activation='relu'))
model.add(keras.layers.Dense(5, activation='softmax'))

In [40]:
# Compile model
tf.random.set_seed(42)
np.random.seed(42)

#Optimizer:
adam = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss='sparse_categorical_crossentropy', optimizer=adam, metrics=['accuracy'])

In [41]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

model.fit(train_x, train_y, validation_data=(test_x, test_y), 
          epochs=100, batch_size=100, callbacks=callback)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 4s 4s/step - accuracy: 0.1341 - loss: 2.4854 - val_accuracy: 0.1111 - val_loss: 1.6781
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 150ms/step - accuracy: 0.1220 - loss: 1.6969 - val_accuracy: 0.4167 - val_loss: 1.3641
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 145ms/step - accuracy: 0.3659 - loss: 1.3670 - val_accuracy: 0.4167 - val_loss: 1.2937
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 122ms/step - accuracy: 0.3537 - loss: 1.2883 - val_accuracy: 0.4167 - val_loss: 1.2862
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 172ms/step - accuracy: 0.3415 - loss: 1.2618 - val_accuracy: 0.4167 - val_loss: 1.2794
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 158ms/step - accuracy: 0.3537 - loss: 1.2344 - val_accuracy: 0.4444 - val_loss: 1.2583
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 127ms/step - accuracy: 0.3659 - loss: 1.2027 - val_accuracy: 0.4722 - val_loss: 1.2295
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 140ms/step - accuracy: 0.4268 - loss: 1.1673 - val_accuracy: 0.4444 - val_

In [42]:
# Test values

test_scores = model.evaluate(train_x, train_y, verbose=0)

test_scores

# In results, first is loss, second is accuracy

[0.46281489729881287, 0.8170731663703918]

In [43]:
# Test values

test_scores = model.evaluate(test_x, test_y, verbose=0)

test_scores

# In results, first is loss, second is accuracy

[0.9446601867675781, 0.5]

In [44]:
# Print the values

print(f"neural network with two hidden layers {model.metrics_names[0]}: {test_scores[0]:.2f}")

print(f"neural network with two hidden layers {model.metrics_names[1]}: {test_scores[1]*100:.2f}%")


neural network with two hidden layers loss: 0.94
neural network with two hidden layers compile_metrics: 50.00%


# LSTM Model

In [45]:
np.random.seed(42)
n_steps = 20
n_inputs = 1

model = keras.models.Sequential([
    
    keras.layers.LSTM(16, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(5, activation='softmax')
])

C:\Users\neera\anaconda3\Lib\site-packages\keras\src\layers\rnn\rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [46]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

In [47]:
tf.random.set_seed(42)
np.random.seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.1)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=60,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 5s 355ms/step - accuracy: 0.2288 - loss: 1.4690 - val_accuracy: 0.5000 - val_loss: 1.8488
Epoch 2/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 53ms/step - accuracy: 0.4570 - loss: 1.6953 - val_accuracy: 0.5000 - val_loss: 1.2644
Epoch 3/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.4375 - loss: 1.4472 - val_accuracy: 0.3056 - val_loss: 1.5274
Epoch 4/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step - accuracy: 0.5149 - loss: 1.1380 - val_accuracy: 0.4722 - val_loss: 1.0745
Epoch 5/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - accuracy: 0.5366 - loss: 1.0723 - val_accuracy: 0.5000 - val_loss: 1.1048
Epoch 6/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - accuracy: 0.4670 - loss: 1.1564 - val_accuracy: 0.6389 - val_loss: 0.9058
Epoch 7/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - accuracy: 0.5427 - loss: 0.9700 - val_accuracy: 0.6111 - val_loss: 0.8945
Epoch 8/60
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - accuracy: 0.5522 - loss: 0.9537 - val_accuracy: 0.6111 - val_loss: 0.8134

In [48]:
# evaluate the model

scores = model.evaluate(train_x, train_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.8846220374107361, 0.6219512224197388]

In [49]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.8619834184646606, 0.6388888955116272]

In [50]:
# extract the accuracy from model.evaluate

print("%s of LSTM with one hidden layer: %.2f" % (model.metrics_names[0], scores[0]))
print("%s of LSTM with one hidden layer: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss of LSTM with one hidden layer: 0.86
compile_metrics of LSTM with one hidden layer: 63.89%


# Deep LSTM Model (with only two layers)

In [51]:
np.random.seed(42)
n_steps = 20
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.LSTM(16, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.LSTM(16),
    keras.layers.Dense(5, activation='sigmoid')
])

In [52]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

In [53]:
tf.random.set_seed(42)
np.random.seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.02)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)


Epoch 1/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 7s 413ms/step - accuracy: 0.2335 - loss: 1.5792 - val_accuracy: 0.5000 - val_loss: 1.1975
Epoch 2/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - accuracy: 0.4353 - loss: 1.3013 - val_accuracy: 0.5278 - val_loss: 1.0785
Epoch 3/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.4770 - loss: 1.1425 - val_accuracy: 0.5556 - val_loss: 1.1546
Epoch 4/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 53ms/step - accuracy: 0.4870 - loss: 1.1426 - val_accuracy: 0.5833 - val_loss: 1.0262
Epoch 5/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.4870 - loss: 1.0208 - val_accuracy: 0.5278 - val_loss: 0.8977
Epoch 6/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step - accuracy: 0.5444 - loss: 0.9087 - val_accuracy: 0.6667 - val_loss: 0.8820
Epoch 7/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.5344 - loss: 1.0280 - val_accuracy: 0.6389 - val_loss: 0.8307
Epoch 8/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 51ms/step - accuracy: 0.5617 - loss: 0.8477 - val_accuracy: 0.6111 - val_loss: 0.8565

In [54]:
# evaluate the model

scores = model.evaluate(train_x, train_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.7207087278366089, 0.6585366129875183]

In [55]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.9497265219688416, 0.6944444179534912]

In [56]:
# extract the accuracy from model.evaluate

print("%s of LSTM with two hidden layer: %.2f" % (model.metrics_names[0], scores[0]))
print("%s of LSTM with two hidden layer : %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss of LSTM with two hidden layer: 0.95
compile_metrics of LSTM with two hidden layer : 69.44%


# GRU Model (with only one layer)

In [57]:
np.random.seed(42)
n_steps = 20
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(32, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(5, activation='softmax')
])

In [58]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

In [59]:
np.random.seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.004)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 5s 363ms/step - accuracy: 0.3418 - loss: 1.4730 - val_accuracy: 0.4444 - val_loss: 1.3201
Epoch 2/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - accuracy: 0.4075 - loss: 1.3609 - val_accuracy: 0.4444 - val_loss: 1.2365
Epoch 3/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - accuracy: 0.4075 - loss: 1.3004 - val_accuracy: 0.4722 - val_loss: 1.1818
Epoch 4/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - accuracy: 0.4453 - loss: 1.2574 - val_accuracy: 0.4722 - val_loss: 1.1431
Epoch 5/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - accuracy: 0.4631 - loss: 1.2209 - val_accuracy: 0.5278 - val_loss: 1.1086
Epoch 6/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.4709 - loss: 1.1822 - val_accuracy: 0.5556 - val_loss: 1.0724
Epoch 7/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - accuracy: 0.4709 - loss: 1.1370 - val_accuracy: 0.5556 - val_loss: 1.0328
Epoch 8/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - accuracy: 0.4888 - loss: 1.0838 - val_accuracy: 0.5556 - val_loss: 0.9899

In [60]:
# evaluate the model

scores = model.evaluate(train_x, train_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.7063842415809631, 0.6219512224197388]

In [61]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.788497805595398, 0.6111111044883728]

In [62]:
# extract the accuracy from model.evaluate

print("%s of GRU  with one hidden layer: %.2f" % (model.metrics_names[0], scores[0]))
print("%s of GRU  with one hidden layer: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss of GRU  with one hidden layer: 0.79
compile_metrics of GRU  with one hidden layer: 61.11%


# Deep GRU Model (with only two layers)

In [63]:
np.random.seed(42)

n_steps = 20
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(16, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.GRU(16),
    keras.layers.Dense(5, activation='softmax')
])

In [64]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

In [65]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 9s 513ms/step - accuracy: 0.3357 - loss: 1.3936 - val_accuracy: 0.4722 - val_loss: 1.2044
Epoch 2/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - accuracy: 0.4788 - loss: 1.2641 - val_accuracy: 0.4722 - val_loss: 1.1304
Epoch 3/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 57ms/step - accuracy: 0.4688 - loss: 1.1716 - val_accuracy: 0.4444 - val_loss: 1.0763
Epoch 4/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 64ms/step - accuracy: 0.4788 - loss: 1.0872 - val_accuracy: 0.4444 - val_loss: 1.0670
Epoch 5/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 59ms/step - accuracy: 0.4888 - loss: 1.0568 - val_accuracy: 0.5000 - val_loss: 1.0428
Epoch 6/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - accuracy: 0.5349 - loss: 1.0356 - val_accuracy: 0.5833 - val_loss: 0.9517
Epoch 7/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.5800 - loss: 0.9626 - val_accuracy: 0.5833 - val_loss: 0.8483
Epoch 8/20
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.5778 - loss: 0.8752 - val_accuracy: 0.5833 - val_loss: 0.8035

In [66]:
# evaluate the model

scores = model.evaluate(train_x, train_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.6894240379333496, 0.6341463327407837]

In [67]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.8110677003860474, 0.6666666865348816]

In [68]:
# extract the accuracy from model.evaluate

print("%s of GRU with two hidden layer: %.2f" % (model.metrics_names[0], scores[0]))
print("%s of GRU with two hidden layer: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss of GRU with two hidden layer: 0.81
compile_metrics of GRU with two hidden layer: 66.67%


# Discussion

## Test values of each model 

Baseline Train Accuracy: 0.34146341463414637
Baseline Test Accuracy: 0.4166666666666667

loss of neural network with one hidden layer : 0.89
compile_metrics of neural network with one hidden layer : 61.11%

neural network with two hidden layers loss: 0.94
neural network with two hidden layers compile_metrics: 50.00%

loss of LSTM with one hidden layer: 0.86
compile_metrics of LSTM with one hidden layer: 63.89%

loss of LSTM with two hidden layer: 0.95
compile_metrics of LSTM with two hidden layer : 69.44%

loss of GRU  with one hidden layer: 0.79
compile_metrics of GRU  with one hidden layer: 61.11%

loss of GRU with two hidden layer: 0.81
compile_metrics of GRU with two hidden layer: 66.67%

## Which model performs is best

The best model among those I have build is the GRU with one hidden layer. It scores the highest accuracy at 69.44%, meaning it makes the most correct predictions. Although its loss isn't the absolute lowest, it's still quite good, suggesting it strikes a nice balance — it neither underfits nor overfits the data. GRUs are also known for being efficient, especially with time series data, making this model not just effective but also practical for many tasks.

## Best Model vs Baseline

The GRU model with one hidden layer significantly outperforms the baseline model. The baseline's accuracy rates are 32.93% for training and 44.44% for testing, which are much lower than the 69.44% accuracy achieved by the GRU model. This jump in accuracy means the GRU model is much better at predicting correctly compared to the baseline. Essentially, while the baseline model struggles to capture and predict the patterns in the data accurately, the GRU model did a reasonably good  job of understanding and forecasting based on the same data. This makes the GRU model a much more reliable choice for making predictions.

# 10-fold cross validation

The dataset is very small. This means my test values are likely unreliable. Using my  best model and running  a 10-fold cross validation on it.



Building 10-fold cross validation model with GRU one layer as it is my best model. 

In [69]:
# Common imports
import numpy as np
import tensorflow as tf
from tensorflow import keras
import pandas as pd

np.random.seed(123)

In [70]:
data = pd.read_csv("movement.csv")

In [71]:
y = data['Target']
x = data.drop('Target', axis=1)

In [72]:
from sklearn.model_selection import TimeSeriesSplit

# Configure the TimeSeriesSplit with 10 splits
tscv = TimeSeriesSplit(n_splits=10)


In [73]:
from sklearn.preprocessing import LabelEncoder


In [74]:
encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)


In [75]:
import numpy as np
x_reshaped = np.reshape(x.values, (x.shape[0], x.shape[1], 1))  # Only reshape if 'x' is not already in this format.


In [76]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import GRU, Dense

# Define the model architecture
model = Sequential([
    GRU(50,input_shape=(x.shape[1], 1)),  # First GRU layer
    Dense(5, activation='softmax')  # Output layer for 5 classes
])


C:\Users\neera\anaconda3\Lib\site-packages\keras\src\layers\rnn\rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [77]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])


In [78]:
import numpy as np

# Reshape x to be 3D [samples, timesteps, features]
x_reshaped = np.reshape(x.values, (x.shape[0], x.shape[1], 1))


In [79]:
from sklearn.utils import shuffle

# Shuffle the data
x_reshaped, y_encoded = shuffle(x_reshaped, y_encoded, random_state=42)

# Now proceed with your cross-validation setup


In [80]:
scores = []
fold_index = 0 
for train_index, test_index in tscv.split(x_reshaped):
    x_train, x_test = x_reshaped[train_index], x_reshaped[test_index]
    y_train, y_test = y_encoded[train_index], y_encoded[test_index]  # Ensure this uses y_encoded and is indexed properly

    # Fit the model
    model.fit(x_train, y_train, epochs=10, batch_size=32, verbose=0)  # Use y_train here, not y_encoded
    # Evaluate the model
    loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
    scores.append(accuracy)
    
    print(f"Accuracy for fold {fold_index}: {accuracy:.2f}%")
    fold_index += 1

# Calculate the mean accuracy
mean_accuracy = np.mean(scores)
print("Mean Accuracy of 10 fold cross validation: {:.2f}%".format(mean_accuracy * 100))


Accuracy for fold 0: 0.40%
Accuracy for fold 1: 0.50%
Accuracy for fold 2: 0.20%
Accuracy for fold 3: 0.60%
Accuracy for fold 4: 0.50%
Accuracy for fold 5: 0.50%
Accuracy for fold 6: 0.60%
Accuracy for fold 7: 0.70%
Accuracy for fold 8: 0.50%
Accuracy for fold 9: 0.50%
Mean Accuracy of 10 fold cross validation: 50.00%



The mean accuracy of 52.00% from the 10-fold cross-validation is lower than the 69.44% accuracy achieved by the best working model, the GRU with one hidden layer.